src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -124,8 +124,14 @@ private AbstractMemorySegmentImpl asSliceNoCheck(long offset, long newSize) {
         if (elementLayout.byteSize() == 0) {
             throw new IllegalArgumentException("Element layout size cannot be zero");
         }
-        if (byteSize() % elementLayout.byteSize() != 0) {
-            throw new IllegalArgumentException("Segment size is no a multiple of layout size");
+        if (elementLayout.byteAlignment() > elementLayout.byteSize()) {
```

Comment:
I wonder if we should add a similar check on the various `MemorySegment/MemoryAddress::[get/set]atIndex` ? I think accessing hyper-aligned elements is always going to fail, for certain offsets - in fact when deriving a var handle from a layout path, it is never possible to have an array var handle for an hyper-aligned element layout.